cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaCluster.java

```diff
@@ -1675,13 +1669,72 @@ private String generatePerBrokerConfiguration(NodeRef node, KafkaPool pool, Map<
                         .withAuthorization(cluster, authorization)
                         .withCruiseControl(cluster, ccMetricsReporter, node.broker())
                         .withUserConfiguration(configuration, node.broker() && ccMetricsReporter != null);
+        buildBrokerConfiguration(node, pool, builder);
+        return builder.build().trim();
+    }
 
-        if (useKRaft) {
-            builder.withKRaft(cluster, namespace, pool.processRoles, nodes());
-        } else {
-            builder.withZookeeper(cluster);
+    /**
+     * Adds ZooKeeper and/or KRaft connection and/or ZooKeeper migration depending on the role of the node (broker or controller)
+     * and the Kafka metadata configuration state.
+     * This method actually implements the following table.
+     *
+     * +----------------+--------------+--------------+
+     * |                | Controller   | Broker       |
+     * +----------------+--------------+--------------+ ---> strimzi.io/kraft = disabled
+     * | ZK             | -            | Zk           |
+     * +----------------+--------------+--------------+ ---> strimzi.io/kraft = migration
+     * | PRE_MIGRATION  | KRaft        | Zk           |
+     * |                | Zk           |              | ---> controllers deployed
+     * |                | Zk-migration |              |
+     * +----------------+--------------+--------------+
+     * | MIGRATION      | KRaft        | KRaft        |
+     * |                | Zk           | Zk           | ---> brokers rolled
+     * |                | Zk-migration | Zk-migration |
+     * +----------------+--------------+--------------+ ---> strimzi.io/kraft = enabled
+     * | POST_MIGRATION | KRaft        | KRaft        |
+     * |                | Zk           |              | ---> brokers rolled
+     * |                | Zk-migration |              |
+     * +----------------+--------------+--------------+
+     * | KRAFT          | KRaft        | KRaft        | ---> controllers rolled
+     * +----------------+--------------+--------------+
+     *
+     * @param node node on which the configuration is applied
+     * @param pool pool hosting the node
+     * @param builder KafkaBrokerConfigurationBuilder instance to use to build the node configuration
+     */
+    private void buildBrokerConfiguration(NodeRef node, KafkaPool pool, KafkaBrokerConfigurationBuilder builder) {
```

Comment:
You call the method `build...` ... but it does not build, it just modifies the builder. That seems weird. I gues you should call it differently or have it build and return the configuration.

Reply:
Regarding the name I would suggest `withZooKeeperOrKRaftConfiguration`. It seems to take the builder pattern with `withXXX` method.